Add remote coinbase transaction handling to the wallet

philipr-za opened the pull request:
## Description

This PR adds the functionality for an external application to request a complete Coinbase transaction from the wallet by providing the reward, fees and block height at which this coinbase will appear. The Transaction Service will then request a coinbase key from the Output Manger Service which will generate a private key from the “coinbase” branch of the current Key Manager using the Block Height as an index. The transaction service will then produce and return the Coinbase transaction to the requester and start monitoring the chain for the Coinbase output to be mined. If the chain advances to a height beyond the Coinbase transactions block height and it has not appeared in the chain it is cancelled.

If multiple requests are made for a coinbase transaction of the same height the previous transaction and outputs are cancelled and the new output and transaction becomes the active Coinbase transaction for that block height.

## How Has This Been Tested?
Tests are provided that test the Output Manager and Transaction manager logic in the database backend and from the Service API perspectives.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

stringhandler commented on base_layer/core/src/mining/mod.rs:
```diff
@@ -20,10 +20,18 @@
+#[cfg(any(feature = "base_node", feature = "transactions"))]
+pub use coinbase_builder::CoinbaseBuildError;
+#[cfg(any(feature = "base_node", feature = "transactions"))]
 pub use coinbase_builder::CoinbaseBuilder;
```
<Suggestion> Maybe this struct should move to the `transactions` mod?

philipr-za replied:
Hmm, ja I think it could. I will see if I can do it neatly.

stringhandler commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -436,6 +449,45 @@ where
+
+        self.resources
+            .db
+            .cancel_pending_transaction_below_block_height(block_height)
```
Are all transactions cancelled or only coinbase transactions?

philipr-za replied:
Any transactions that have a blockheight specified which will only be coinbase transactions. Perhaps that column should be explicitly `coinbase_block_height` so as not to confuse with other types of maturity that might come later.

stringhandler commented on base_layer/wallet/src/output_manager_service/storage/database.rs:
```diff
@@ -519,6 +525,18 @@ where T: OutputManagerBackend + Clone + 'static
             .and_then(|inner_result| inner_result)
     }
 
+    pub async fn cancel_pending_transaction_below_block_height(
```
This method name mismatches the one actually called, where it only cancels __at__ the block height. Is there something going to be added in future?

philipr-za replied:
Oh I actually moved from this to just `at_block_height` because I realized it should only relate to the current block_height and let the monitoring protocols handle the ones at lower blockheight. Will update this.

stringhandler reviewed (commented):
Looking good. Just some questions

philipr-za committed:
Add remote coinbase transaction handling to the wallet

This PR adds the functionality for an external application to request a complete Coinbase transaction from the wallet by providing the reward, fees and block height at which this coinbase will appear. The Transaction Service will then request a coinbase key from the Output Manger Service which will generate a private key from the “coinbase” branch of the current Key Manager using the Block Height as an index. The transaction service will then produce and return the Coinbase transaction to the requester and start monitoring the chain for the Coinbase output to be mined. If the chain advances to a height beyond the Coinbase transactions block height and it has not appeared in the chain it is cancelled.
If multiple requests are made for a coinbase transaction of the same height the previous transaction and outputs are cancelled and the new output and transaction becomes the active Coinbase transaction for that block height.

stringhandler reviewed (approved):
Did a code review, but didn't test it locally. Looks fine. Any problems should come out when we start calling this from the miner code. The missing transactions are logged in another issue, as mentioned
